### ISSUE TITLE
Incorrect waitrequest handling in Avalon MM BFM

### ISSUE THREAD

**Issue by aghoras** [UVVM/UVVM#241]
Hello All,
I believe I have found a bug with the Avalon MM BFM:
If the Avalon MM Slave uses *waitrequest* and does not have it asserted when the read signal from the master goes high, the avalon_mm_read_response procedure (and avalon_mm_check_response procedure) will sample the readdata bus before the rising edge of the clock where the slave is expected to see the read signal asserted which is obviously too soon.  

UVVM does the following:

1. 	BFM asserts the read signal mid clock cycle
2. 	On the next rising edge of the clock (at the cursor), BFM samples waitrequest as 0
        a) Since waitrequest is 0, the avalon_mm_read_request completes and avalon_mm_read_response  samples the readdata bus
3. BFM clears the read signal

This does not give the slave any time to output the readdata bus. When waitrequest is not asserted during the read cycle, the correct behavior should be:
1. BFM asserts the read signal mid clock cycle
2. On the next rising edge of the clock (at the cursor), BFM samples waitrequest as 0
       a. Since waitrequest is 0, the avalon_mm_read_request waits for the next rising edge before returning and avalon_mm_read_response  samples the readdata bus
3.  BFM clears the read signal

Here a simulation of the existing Avalon BFM behavior:
![image](https://github.com/UVVM/UVVM/assets/3997717/b3e9a6d9-9572-450c-a9d7-971b8bbecd93)

Here is what should happen:
![image](https://github.com/UVVM/UVVM/assets/3997717/4f08e121-53ec-4f69-ae88-97c6bb95b195)

	1. The host asserts read (point a) and drives the address bus
	2. These two steps happen simultaneously at point (b)
		a. The host samples the waitrequest line and since it's 0, it clears read.
		b. The agent samples the read line and since it's 1, it drive the data on the readdata bus
       3. The host samples the readdata bus a point c).

The following patch fixes the problem. I will create a pull request as well:
```
diff --git a/bitvis_vip_avalon_mm/src/avalon_mm_bfm_pkg.vhd b/bitvis_vip_avalon_mm/src/avalon_mm_bfm_pkg.vhd
index c2e703c..5cb7f73 100644
--- a/bitvis_vip_avalon_mm/src/avalon_mm_bfm_pkg.vhd
+++ b/bitvis_vip_avalon_mm/src/avalon_mm_bfm_pkg.vhd
@@ -510,6 +510,8 @@ package body avalon_mm_bfm_pkg is
     constant local_proc_call   : string                                                     := local_proc_name & "(A:" & to_string(addr_value, HEX, AS_IS, INCL_RADIX) & ")";
     variable timeout           : boolean                                                    := false;
     variable v_proc_call       : line;  -- Current proc_call, external or local
+    -- set to false if wait request is used but it is not asserted on the first rising edge of the clock after read is asserted
+    variable v_add_wait_request_delay : boolean                                               := false; 
     variable v_normalized_addr : std_logic_vector(avalon_mm_if.address'length - 1 downto 0) := normalize_and_check(std_logic_vector(addr_value), avalon_mm_if.address, ALLOW_NARROWER, "addr", "avalon_mm_if.address", msg);
 
     variable v_time_of_rising_edge  : time := -1 ns; -- time stamp for clk period checking
@@ -559,6 +561,9 @@ package body avalon_mm_bfm_pkg is
         if is_waitrequest_active(avalon_mm_if, config) then
           wait until rising_edge(clk);
         else
+          if cycle = 1 then
+            v_add_wait_request_delay := true;
+          end if;
           exit;
         end if;
         if cycle = config.max_wait_cycles then
@@ -578,7 +583,13 @@ package body avalon_mm_bfm_pkg is
     end if;
 
     avalon_mm_if <= init_avalon_mm_if_signals(avalon_mm_if.address'length, avalon_mm_if.writedata'length, avalon_mm_if.lock) after v_clock_period / 4;
-
+    -- if wait request is not asserted on the first cycle, wait until the next
+    -- rising edge until data becomes available by the agent
+    if v_add_wait_request_delay then
+        log(config.id_for_bfm, v_proc_call.all & " inserting one cycle delay for agent to drive readdata"& add_msg_delimiter(msg), scope, msg_id_panel);
+        wait until rising_edge(clk);
+    end if;
+          
     if ext_proc_call = "" then
       log(config.id_for_bfm, v_proc_call.all & " completed. " & add_msg_delimiter(msg), scope, msg_id_panel);
     else

```

**Comment by erick166** [comment]
Hi,
It seems correct what you are reporting. However, the fix that you made might cause some issues when using pipelined transactions. It might be better to add the wait in the read response, can you test by adding this wait statement at the beginning of avalon_mm_read_response()?

![image](https://github.com/UVVM/UVVM/assets/73584022/7435a21d-34d0-4c0d-a434-7839712d5a3d)

**Comment by aghoras** [comment]
I can try that.. but I'm afraid of adding an extra clock cycle to the case where waitrequest is asserted prior to read. Do you think that will be a problem?

**Comment by erick166** [comment]
True, the extra cycle might not be compliant with the timing of the interface defined in the specification.
Then I think we can use your patch for this bug and add an extra fix for the pipeline problem in #233.

**Comment by UVVM** [comment]
This issue has been fixed in the latest release v2 2024.07.03

**Comment by aghoras** [comment]
I hate to do this but I owe it to the community:

When I filed the original bug, I had checked with Intel regarding their Avalon MM wait request behavior. I started down this path because Intel's FIFO BFM was not working with the UVVM Avalon Memory map master.

Response from Intel:
1. The agent (or slave) is supposed to assert the waitrequest signal asynchronously when it detects the read signal as asserted if it cannot accept data or needs more time to prepare the data. This assertion can happen immediately upon the read request without waiting for a clock edge.

2. If the agent does not need to assert waitrequest for the current operation, such as when reading from a non-empty FIFO, it is not required to drive read data before the next rising clock edge. The agent should drive readdata synchronously with the clock, such that the data is available on the readdata bus at the next rising edge after the read request.

3. The behavior you've observed with the Altera Avalon MM FIFO BFM, where it clears waitrequest when the FIFO is not empty and asserts it when the FIFO is drained, is consistent with synchronous behavior. The readdata is driven after the first rising edge of the clock following the assertion of the read signal, which is the expected behavior for a synchronous module.

4. If the agent does not need to assert waitrequest, the expected behavior would be: a. The host asserts read and drives the address at point a b. The agent drives the data on readdata at the next rising clock edge (point b) c. The host samples readdata at the subsequent rising clock edge (point c) and lowers read d. The agent ignores the read signal at point c since the transaction is complete


Conclusion After Testing With Hardware
Intel's answer is wrong: The fabric does not wait an extra cycle before sampling the readdata bus. In the figure below, the NIOS V sees a value of 0x1 and not 0x2 

![image](https://github.com/user-attachments/assets/dda374c7-bc26-43c5-8d5f-0e98ddb2fdc3)


**The UVVM behavior was correct and the salve should either asynchronously drive waitrequest or drive readdata before the next rising edge after the read signal is asserted.** This figure from the Avalon Memory Mapped Interface Specification explains the transaction a little better.

![image](https://github.com/user-attachments/assets/8506da84-f8c9-4074-ad59-9b553f1bde93)



The correct transaction without waitrequest is
![image](https://github.com/user-attachments/assets/201b1476-b2d5-4bc1-84b4-5d05ea2af704)



Master drives address and read at a
The slave drives the data on read_data before the next rising edge of the clock
Master samples the data at c
A burst read would look like:
![image](https://github.com/user-attachments/assets/7e8f7836-29cd-4118-9022-def2a0d3cc7a)




A read bust from signal tap:
![image](https://github.com/user-attachments/assets/93d2fbee-db66-4aa2-abf6-acac054f8c15)

_**As a result, I need back out this fix from the baseline**_

**Comment by UVVM** [comment]
We already included this, so we now need to check whether to revert this.
Comments are welcome.

**Comment by UVVM** [comment]
Hi Aghoras, 
We have decided to leave the current implementation as is for now to avoid issues with backwards-compatibility, but will update the documentation to clarify the timing behavior of the UVVM implementation. 
Since the Avalon MM BFM currently does not support the combination of waitrequest and fixed read-latency (using _num_wait_states_read_), we expect that _waitrequest_ will normally be used in combination with _readdatavalid_, in which case the Avalon MM standard dictates that there must be at least one cycle of latency between assertion of _read_ and assertion of _readdatavalid_. Therefore, the delay added by the pull request should not introduce any risk of missing the _readdatavalid_ signal, since the delay is only added when _waitrequest_ is deasserted at the first rising edge after _read_ is asserted. Of course, feel free to comment if you disagree with this reasoning. 

That being said, we will create an internal issue regarding the need to support waitrequest in combination with fixed read-latency.
